src/rust/engine/process_execution/sandboxer/src/lib.rs

```diff
@@ -0,0 +1,226 @@
+        self.do_serve(false).await
+    }
+
+    pub async fn do_serve(
+        &mut self,
+        with_shutdown_thread: bool,
+    ) -> Result<(), Box<dyn std::error::Error>> {
+        // `bind()` expects the socket file not to exist, so we ensure that.
+        ensure_socket_file_removed(&self.socket_path)?;
+        // Wait a beat, to let any previous sandboxer process notice this removal and exit.
+        // There's no great harm in leaving a stray sandboxer process running, it will
+        // receive no requests (since the socket file was pulled out from under it) so eventually
+        // it'll shut down due to idleness. But it's nicer if we can let it go right away.
+        tokio::time::sleep(SandboxerService::POLLING_INTERVAL).await;
+
+        ensure_parent_dir(&self.socket_path)?;
+        let listener = UnixListener::bind(&self.socket_path)?;
+        // `bind()` creates the socket file, so now we can poll it, in case the
+        // user deletes it.
+        // NB: We can't easily use the notify crate to watch changes to the socket file,
+        //  because on MacOS the FSEvent watches are path-based, not inode-based, and they
+        //  are batched. So we may get a spurious notification for the removal above, and not
+        //  know if we should act on it. We could use notify's PollWatcher, but it's easier to
+        // implement simple polling ourselves, without the overkill of the notify crate.
+        let socket_path = self.socket_path.clone();
+        let inactivity_counter = Arc::new(AtomicUsize::new(0));
+        let inactivity_counter_ref = inactivity_counter.clone();
+        if with_shutdown_thread {
+            thread::spawn(move || {
+                loop {
+                    // Note: just a regular thread::sleep() since this is not an async context.
+                    thread::sleep(SandboxerService::POLLING_INTERVAL);
+                    if !fs::exists(&socket_path).unwrap_or(false) {
+                        warn!("Socket file {} deleted. Exiting.", &socket_path.display());
+                        process::exit(22);
+                    }
+                    if inactivity_counter_ref.fetch_add(1, Ordering::Relaxed)
+                        > Self::ALLOWED_IDLE_BEATS
+                    {
+                        warn!("Idle time limit exceeded. Exiting.");
+                        process::exit(23);
+                    }
+                }
+            });
+        }
+
+        info!(
+            "Sandboxer server started on socket path {}",
+            self.socket_path.display()
+        );
+        let store = self
+            .store_cli_opt
+            .create_store(task_executor::Executor::new())
+            .await?;
+        let router = Server::builder().add_service(SandboxerGrpcServer::new(SandboxerGrpcImpl {
+            store,
+            inactivity_counter,
+        }));
+        router
+            .serve_with_incoming(UnixListenerStream::new(listener))
```

Comment:
Consider using `.serve_with_incoming_shutdown` here instead. The shutdown thread would then just signal a one-shot channel passed to `.serve_with_incoming_shutdown` and the gRPC router would shutdown cleanly.

Reply:
Oooh, that is much nicer. It means I can get rid of `with_shutdown_thread` and `serve_in_process` since we no long call `process::exit()`. Done.